trace_decoder/src/compact/compact_to_partial_trie.rs

```diff
@@ -8,183 +11,261 @@ use mpt_trie::{
-    let c_hash = hash(&c_bytes);
-    output.code.insert(c_hash, c_bytes);
+impl CompactToPartialTrieExtractionOutput for StateTrieExtractionOutput {
+    fn process_code(&mut self, c_bytes: Vec<u8>) -> CompactParsingResult<()> {
+        let c_hash = hash(&c_bytes);
+        self.code.insert(c_hash, c_bytes);
 
-    Ok(())
-}
+        Ok(())
+    }
 
-fn process_empty() -> CompactParsingResult<()> {
-    // Nothing to do.
-    Ok(())
+    fn process_leaf(
+        &mut self,
+        curr_key: Nibbles,
+        leaf_key: &Nibbles,
+        leaf_node_data: &LeafNodeData,
+    ) -> CompactParsingResult<()> {
+        process_leaf_common(
+            &mut self.trie,
+            curr_key,
+            leaf_key,
+            leaf_node_data,
+            |acc_data, full_k| {
+                Ok(
+                    convert_account_node_data_to_rlp_bytes_and_add_any_code_to_lookup(
+                        acc_data,
+                        full_k,
+                        &mut self.code,
+                        &mut self.storage_tries,
+                    ),
+                )
+            },
+        )
+    }
+    fn get_trie(&mut self) -> &mut HashedPartialTrie {
+        &mut self.trie
+    }
+}
+#[derive(Debug, Default)]
+pub(super) struct StorageTrieExtractionOutput {
+    pub(super) trie: HashedPartialTrie,
 }
 
-fn process_hash(
-    curr_key: Nibbles,
-    hash: TrieRootHash,
-    p_trie: &mut HashedPartialTrie,
-) -> CompactParsingResult<()> {
-    // If we see a hash node at this stage, it must be a hashed out node in the
-    // trie.
-    p_trie.insert(curr_key, hash);
+impl CompactToPartialTrieExtractionOutput for StorageTrieExtractionOutput {
+    fn process_code(&mut self, c_bytes: Vec<u8>) -> CompactParsingResult<()> {
+        Err(CompactParsingError::UnexpectedNodeForTrieType(
+            UnexpectedCompactNodeType::Code,
+            TrieType::Storage,
+        ))
+    }
 
-    Ok(())
+    fn process_leaf(
+        &mut self,
+        curr_key: Nibbles,
+        leaf_key: &Nibbles,
+        leaf_node_data: &LeafNodeData,
+    ) -> CompactParsingResult<()> {
+        process_leaf_common(
+            &mut self.trie,
+            curr_key,
+            leaf_key,
+            leaf_node_data,
+            |_, _| {
+                Err(CompactParsingError::UnexpectedNodeForTrieType(
+                    UnexpectedCompactNodeType::AccountLeaf,
+                    TrieType::Storage,
+                ))
+            },
+        )
+    }
+
+    fn get_trie(&mut self) -> &mut HashedPartialTrie {
+        &mut self.trie
+    }
 }
 
-fn process_leaf(
+fn process_leaf_common<F: FnMut(&AccountNodeData, &Nibbles) -> CompactParsingResult<Vec<u8>>>(
+    trie: &mut HashedPartialTrie,
     curr_key: Nibbles,
     leaf_key: &Nibbles,
     leaf_node_data: &LeafNodeData,
-    output: &mut CompactToPartialTrieExtractionOutput,
+    mut account_leaf_proc_f: F,
 ) -> CompactParsingResult<()> {
     let full_k = curr_key.merge_nibbles(leaf_key);
```

Comment:
This is more of a general concern than specifically for this code: `merge_nibbles` panics on invalid input but that is not obvious from looking at the signature here. where instead it seems like we'd get an error when something's amiss. I think `merge_nibbles` should probably not panic and instead return an error. At the very least we should document the panicking behaviour here.

Reply:
Yeah... This might be worth revisiting as a separate PR. I've opened an issue (#94 ) to discuss this a bit more.